ethereum/forge-test/Bridge.t.sol

```diff
@@ -56,6 +146,1289 @@ contract TestBridge is Test {
+            bytes32(
+                0x0000000000000000000000000000000000000000000000000000000000000000
+            )
+        );
+        bridge.registerChain(vaa);
+        assertEq(
+            bridge.bridgeContracts(testForeignChainId),
+            testForeignBridgeContract
+        );
+    }
+
+    function testShouldAcceptAValidUpgrade() public {
+        MockBridgeImplementation mock = new MockBridgeImplementation();
+        bytes memory data = abi.encodePacked(
+            tokenBridgeModule,
+            actionContractUpgrade,
+            testChainId,
+            addressToBytes32(address(mock))
+        );
+        bytes memory vaa = signAndEncodeVM(
+            1,
+            1,
+            governanceChainId,
+            governanceContract,
+            0,
+            data,
+            uint256Array(testGuardian),
+            0,
+            0
+        );
+
+        bytes32 IMPLEMENTATION_STORAGE_SLOT = 0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc;
+        assertEq(
+            vm.load(address(bridge), IMPLEMENTATION_STORAGE_SLOT),
+            addressToBytes32(address(bridgeImpl))
+        );
+        bridge.upgrade(vaa);
+        assertEq(
+            vm.load(address(bridge), IMPLEMENTATION_STORAGE_SLOT),
+            addressToBytes32(address(mock))
+        );
+
+        assertTrue(
+            MockBridgeImplementation(payable(address(bridge)))
+                .testNewImplementationActive(),
+            "implementation not active"
+        );
+    }
+
+    function testBridgedTokensShouldOnlyBeMintAndBurnableByOwner() public {
+        address owner = address(this);
+        address notOwner = address(0x1);
+        tokenImpl.initialize("TestToken", "TT", 18, 0, owner, 0, 0x0);
+        tokenImpl.mint(owner, 10);
+        tokenImpl.burn(owner, 5);
+
+        vm.expectRevert("caller is not the owner");
+        vm.prank(address(notOwner));
+        tokenImpl.mint(owner, 10);
+
+        vm.expectRevert("caller is not the owner");
+        vm.prank(address(notOwner));
+        tokenImpl.burn(owner, 5);
+    }
+
+    event LogMessagePublished(
+        address indexed sender,
+        uint64 sequence,
+        uint32 nonce,
+        bytes payload,
+        uint8 consistencyLevel
+    );
+
+    function testShouldAttestATokenCorrectly() public {
+        tokenImpl.initialize("TestToken", "TT", 18, 0, address(this), 0, 0x0);
+        bytes memory attestPayload = abi.encodePacked(
+            uint8(2),
+            addressToBytes32(address(tokenImpl)),
+            testChainId,
+            // decimals
+            uint8(18),
+            // symbol (TT)
+            bytes32(
+                0x5454000000000000000000000000000000000000000000000000000000000000
+            ),
+            // name (TestToken)
+            bytes32(
+                0x54657374546f6b656e0000000000000000000000000000000000000000000000
+            )
+        );
+        vm.expectEmit();
```

Comment:
Might be better to fetch the raw message and decode the data into the Attested struct.

Reply:
Agree this should be addressed with #4090